### PR TITLE
MM-8823: Checks for 'read_channel' permission on search results.

### PR DESCRIPTION
**Pull request by mkraft**
#### Summary
When searching, if user does not have system-level or team-level `read_channel` permission then this PR does some extra checks to ensure that search results are only returned for posts that are in channels in which the user has `read_channel` permission (at the channel level).

Nothing really changes for users with `read_channel` at the system or team level.

Essentially this leverages the existing 'in:channel-name' filtering for ElasticSearch, Postgres, and MySQL.

If based on the query and the users permissions the user can't read any channels then their search is set to an empty string.

Here's the query in MySql and Postgres to get the Channels with `read_channel` permission (at the channel context).

New query added:
```sql
-- postgres
explain select channels.* from (
    select userid, channelid, unnest(string_to_array(roles, ' ')) as rolename from channelmembers
) x
left join channels on channelid = channels.id
left join roles on roles.name = x.rolename
where string_to_array(roles.permissions, ' ') && '{read_channel}'::text[]
and userid = 'jwedp8tbubn68mzbf3ufw1xczo';
```
```
                                                  QUERY PLAN
--------------------------------------------------------------------------------------------------------------
 Nested Loop Left Join  (cost=15.69..29.29 rows=1 width=282)
   ->  Hash Left Join  (cost=15.42..28.70 rows=1 width=27)
         Hash Cond: ((unnest(string_to_array((channelmembers.roles)::text, ' '::text))) = (roles.name)::text)
         Filter: (string_to_array((roles.permissions)::text, ' '::text) && '{read_channel}'::text[])
         ->  Bitmap Heap Scan on channelmembers  (cost=4.30..12.32 rows=300 width=69)
               Recheck Cond: ((userid)::text = 'jwedp8tbubn68mzbf3ufw1xczo'::text)
               ->  Bitmap Index Scan on idx_channelmembers_user_id  (cost=0.00..4.29 rows=3 width=0)
                     Index Cond: ((userid)::text = 'jwedp8tbubn68mzbf3ufw1xczo'::text)
         ->  Hash  (cost=10.50..10.50 rows=50 width=662)
               ->  Seq Scan on roles  (cost=0.00..10.50 rows=50 width=662)
   ->  Index Scan using channels_pkey on channels  (cost=0.27..0.58 rows=1 width=282)
         Index Cond: ((channelmembers.channelid)::text = (id)::text)
(12 rows)
```

```sql
-- mysql
explain select Channels.* from (
select
ChannelId, UserId, trim(substring_index(substring_index(B.Roles, ' ', ns.n), ' ', -1)) as rolename
from (
select 1 as n union all select 2 union all
select 3 union all
select 4 union all
select 5 union all
select 6 union all
select 7 union all
select 8 union all
select 9 union all
select 10 union all
select 11
) ns
inner join ChannelMembers B ON ns.n <= char_length(B.Roles) - char_length(replace(B.Roles, ' ', '')) + 1
) x
left join Channels on ChannelId = Channels.Id
left join Roles on Roles.name = x.rolename
where Roles.Permissions REGEXP '[[:<:]]read_channel[[:>:]]'
and UserId = 'jwedp8tbubn68mzbf3ufw1xczo';
```

```
+----+-------------+------------+------------+--------+----------------------------+----------------------------+---------+-----------------------------+------+----------+----------------------------------------------------+
| id | select_type | table      | partitions | type   | possible_keys              | key                        | key_len | ref                         | rows | filtered | Extra                                              |
+----+-------------+------------+------------+--------+----------------------------+----------------------------+---------+-----------------------------+------+----------+----------------------------------------------------+
|  1 | PRIMARY     | B          | NULL       | ref    | idx_channelmembers_user_id | idx_channelmembers_user_id | 106     | const                       |    1 |   100.00 | NULL                                               |
|  1 | PRIMARY     | Channels   | NULL       | eq_ref | PRIMARY                    | PRIMARY                    | 106     | mattermost_test.B.ChannelId |    1 |   100.00 | NULL                                               |
|  1 | PRIMARY     | <derived3> | NULL       | ALL    | NULL                       | NULL                       | NULL    | NULL                        |   11 |    33.33 | Using where; Using join buffer (Block Nested Loop) |
|  1 | PRIMARY     | Roles      | NULL       | ref    | Name                       | Name                       | 259     | func                        |    1 |   100.00 | Using index condition; Using where                 |
|  3 | DERIVED     | NULL       | NULL       | NULL   | NULL                       | NULL                       | NULL    | NULL                        | NULL |     NULL | No tables used                                     |
|  4 | UNION       | NULL       | NULL       | NULL   | NULL                       | NULL                       | NULL    | NULL                        | NULL |     NULL | No tables used                                     |
|  5 | UNION       | NULL       | NULL       | NULL   | NULL                       | NULL                       | NULL    | NULL                        | NULL |     NULL | No tables used                                     |
|  6 | UNION       | NULL       | NULL       | NULL   | NULL                       | NULL                       | NULL    | NULL                        | NULL |     NULL | No tables used                                     |
|  7 | UNION       | NULL       | NULL       | NULL   | NULL                       | NULL                       | NULL    | NULL                        | NULL |     NULL | No tables used                                     |
|  8 | UNION       | NULL       | NULL       | NULL   | NULL                       | NULL                       | NULL    | NULL                        | NULL |     NULL | No tables used                                     |
|  9 | UNION       | NULL       | NULL       | NULL   | NULL                       | NULL                       | NULL    | NULL                        | NULL |     NULL | No tables used                                     |
| 10 | UNION       | NULL       | NULL       | NULL   | NULL                       | NULL                       | NULL    | NULL                        | NULL |     NULL | No tables used                                     |
| 11 | UNION       | NULL       | NULL       | NULL   | NULL                       | NULL                       | NULL    | NULL                        | NULL |     NULL | No tables used                                     |
| 12 | UNION       | NULL       | NULL       | NULL   | NULL                       | NULL                       | NULL    | NULL                        | NULL |     NULL | No tables used                                     |
| 13 | UNION       | NULL       | NULL       | NULL   | NULL                       | NULL                       | NULL    | NULL                        | NULL |     NULL | No tables used                                     |
+----+-------------+------------+------------+--------+----------------------------+----------------------------+---------+-----------------------------+------+----------+----------------------------------------------------+
15 rows in set, 1 warning (0.00 sec)
```

Obviously this will be slower than the existing search because some extra checks need to be made.

#### Ticket Link
[MM-8823](https://mattermost.atlassian.net/browse/MM-8823)

#### Checklist
- [x] Added or updated unit tests (required for all new features)

### CODE REVIEW

File: api4/post.go [hunk 1]
```diff
@@ -300,6 +301,13 @@ func searchPosts(c *Context, w http.ResponseWriter, r *http.Request) {
 
 	posts, err := c.App.SearchPostsInTeam(terms, c.Session.UserId, c.Params.TeamId, isOrSearch)
 
+	for postID, post := range posts.Posts {
+		if !c.App.SessionHasPermissionToChannel(c.Session, post.ChannelId, model.PERMISSION_READ_CHANNEL) {
+			posts.RemovePost(postID)
+			log4go.Debug("Removed post '%v' from search results because of lack of '%v' permission for channel '%v'.", postID, model.PERMISSION_READ_CHANNEL.Id, post.ChannelId)
```
Comment (mkraft) [inline api4/post.go]: Is this useful for debugging? If not I'll remove.

File: api4/post.go [hunk 2]
```diff
@@ -300,6 +301,13 @@ func searchPosts(c *Context, w http.ResponseWriter, r *http.Request) {
 
 	posts, err := c.App.SearchPostsInTeam(terms, c.Session.UserId, c.Params.TeamId, isOrSearch)
 
+	for postID, post := range posts.Posts {
+		if !c.App.SessionHasPermissionToChannel(c.Session, post.ChannelId, model.PERMISSION_READ_CHANNEL) {
```
Comment (jespino) [inline api4/post.go]: The problem with this is that you are removing results from the resolved query, this means that if you want 100 results, you can get 70 or 40, or 0, when you actually have valid results. We can try to do a while over it, until we fill the block of results, but if we include pagination after that, we have to worry about the offset to, could be a pain in performance.

Comment (jespino) [reply api4/post.go]: I was thinking in something more "SQL based", but I'm 0/5 about if that is a good solution, because we can generate a very complex query to express the permissions match.

Comment (mkraft) [reply api4/post.go]: @jespino I think you are correct because the [elasticsearch query limits the results to 100](https://github.com/mattermost/enterprise/blob/advanced-permissions-phase-1/elasticsearch/elasticsearch.go#L255). Looks like we will be force to go more complicated.

Comment (jespino) [reply api4/post.go]: Yes, we can make a simple approach with a while loop, and check how it performs, and how we can improve. Other option is solve it inside mysql and elasticsearch.

**Comment by jespino** [comment]
Still worried about the performance of this, because you are getting the list of accessible channels and then pass it to the search engine, but actually that is the current behavior, so I think must be acceptable. Anyway I want to test it agaist a big database to see the performance. Because a huge channels table, could still have a poor performance.

**Comment by mkraft** [comment]
@jespino Here are some search test results run on my local machine with 1,200 channels and 5,000 users:

branch | `teamId` in `where` clause | DB | Post count | SampleA (s) | SampleB (s) | SampleC (s) | Average (ms) | Diff (%)
-- | -- | -- | -- | -- | -- | -- | -- | --
MM-8823 | no | mysql | 1,488,096 | 1.705 | 1.717 | 1.677 | 1700 | 3%
MM-8826 | yes | mysql | 1,488,096 | 1.643 | 1.432 | 1.740 | 1605 | -2%
master | n/a | mysql | 1,488,096 | 1.606 | 1.655 | 1.667 | 1642 |  0%

**Comment by jasonblais** [comment]
@mkraft Do you have sample cases where the user can search for posts from a channel they don't belong to?

I wasn't able to via the UI, and the API requires the `read_channel` permission, but I didn't cover all the cases of course,

**Comment by mkraft** [comment]
@jasonblais Great callout. I have added that test scenario with a change to the existing search query. I have marked the PR as 'Awaiting Submitter Action' as I'll be redoing the perf tests. I'll ping you when done.

**Comment by mkraft** [comment]
@jasonblais Previously this PR was simply *limiting* the search results to a subset of membership-based results (intersect of membership results ∩ results in channels with `read_channel` permission). 

If we're going to *expand* results to include non-membership results I'll need to know which groups a user should see search results for from these types:
- open channels
- private channels
- direct channels
- group channels

**Comment by mkraft** [comment]
@jasonblais Just a note that based on https://gist.github.com/mkraft/17f62c66802bf2d68806586ecede97bd the scenario of searching for posts without membership is no longer required.

File: api4/post.go [hunk 3]
```diff
@@ -316,6 +338,54 @@ func searchPosts(c *Context, w http.ResponseWriter, r *http.Request) {
 	w.Write([]byte(c.App.PostListWithProxyAddedToImageURLs(posts).ToJson()))
 }
 
+func hasTeamOrSystemContextPermission(teamID string, userID string, permission *model.Permission, app *app.App) bool {
```
Comment (jwilander) [inline api4/post.go]: Might be worth throwing this function and the two below into the `app` package (intersect() would probably be better in `utils` actually). That way we can keep the api packages lighter and I can definitely see potential for these functions to be re-used by other packages in the future, similar to the other permission related functions

File: api4/post.go [hunk 4]
```diff
@@ -316,6 +338,54 @@ func searchPosts(c *Context, w http.ResponseWriter, r *http.Request) {
...
+		} else {
+			intersectChannelNames = channelNames
+		}
+		searchParam.InChannels = intersectChannelNames
+	}
+
+	// If intersect of channels is empty clear the search params
+	// because there are no channels user can read (for give search).
+	if len(intersectChannelNames) == 0 {
+		searchParams = model.ParseSearchParams("")
```
Comment (jwilander) [inline api4/post.go]: The user might not have any read_channel permissions for channels on the team but I think we should still search DMs and GMs, right? So we could probably just strip the `InChannel` bit out and return the rest of the search params instead of blanking it

File: store/sqlstore/channel_store.go [hunk 5]
```diff
@@ -1469,3 +1471,61 @@ func (s SqlChannelStore) GetMembersByIds(channelId string, userIds []string) sto
...
+	})
+}
+
+func getChannelsWithChannelContextPermissionMysqlQuery(permissionName string, roleCount int) string {
+	query := `
+			select Channels.* from (
+			select
+			ChannelId, UserId, trim(substring_index(substring_index(B.Roles, ' ', ns.n), ' ', -1)) as rolename
+			from (
+			select 1 as n union all `
```
Comment (jwilander) [inline store/sqlstore/channel_store.go]: Can we use the same capitalization we do for other queries? e.g. 'SELECT Channels.* From ...`

File: store/sqlstore/channel_store.go [hunk 6]
```diff
@@ -1469,3 +1471,61 @@ func (s SqlChannelStore) GetMembersByIds(channelId string, userIds []string) sto
...
+	}
+
+	query += `select ` + fmt.Sprintf("%v", roleCount) + `
+			) ns
+			inner join ChannelMembers B ON ns.n <= char_length(B.Roles) - char_length(replace(B.Roles, ' ', '')) + 1
+			) x
+			left join Channels on ChannelId = Channels.Id
+			left join Roles on Roles.name = x.rolename
+			where Roles.Permissions REGEXP '[[:<:]]` + permissionName + `[[:>:]]'
+			and UserId = :UserId`
```
Comment (jwilander) [inline store/sqlstore/channel_store.go]: This is quite the query, mind giving a brief overview of what it's doing?

File: store/sqlstore/channel_store.go [hunk 7]
```diff
@@ -1469,3 +1471,61 @@ func (s SqlChannelStore) GetMembersByIds(channelId string, userIds []string) sto
 		}
 	})
 }
+
+// GetChannelsWithChannelContextPermission retrieves the channels for a given user with an associated channel-context of a permission.
+func (s SqlChannelStore) GetChannelsWithChannelContextPermission(userID string, permissionName string) store.StoreChannel {
```
Comment (jwilander) [inline store/sqlstore/channel_store.go]: Can you add a store unit test for this function?

File: store/sqlstore/channel_store.go [hunk 8]
```diff
@@ -30,6 +30,8 @@ const (
 	CHANNEL_MEMBERS_COUNTS_CACHE_SEC  = 1800 // 30 mins
 
 	CHANNEL_CACHE_SEC = 900 // 15 mins
+
+	TOTAL_ROLE_COUNT = 11
```
Comment (jwilander) [inline store/sqlstore/channel_store.go]: Why 11?

Comment (mkraft) [reply store/sqlstore/channel_store.go]: That's the [number of roles](https://github.com/mattermost/mattermost-server/pull/8232/files#diff-6a93e94fc8adb70b1a7468d9e2324560R172) for Phase 1 permissions. I had considered invoking `MakeDefaultRoles()` and then counting the result but it seemed like overkill to do each query, but I'm not totally opposed to it.

Phase 3 will introduce custom roles and that's when this count would need to be a query.

Comment (jwilander) [reply store/sqlstore/channel_store.go]: Ok cool, if it's going to be fixed later anyways that's fine

**Comment by jasonblais** [comment]
@mkraft Do we need this for v4.9 as part of Phase 1 permissions?

**Comment by mkraft** [comment]
@jasonblais I'm thinking this can wait until the next phase with the side effect being that the read_channel permission won't really do anything yet.

@grundleborg @jespino Do you see any issues with that?

**Comment by jespino** [comment]
I agree with martin, I think is better to wait until Phase 2, because It have some extra implications that have to be evaluated carefully (compliance, performance...).

File: api4/post.go (hunk 4 above)
Comment (mkraft) [reply api4/post.go]: I'm having trouble understanding this comment. Could you rephrase?

File: store/sqlstore/channel_store.go (hunk 6 above)
Comment (mkraft) [reply store/sqlstore/channel_store.go]: @jwilander I think the easiest way to understand it is to look at the Postgres query:

```sql
SELECT channels.* FROM (
    SELECT userid, channelid, UNNEST(STRING_TO_ARRAY(roles, ' ')) AS rolename FROM channelmembers
) x
LEFT JOIN channels ON channelid = channels.id
LEFT JOIN roles ON roles.name = x.rolename
WHERE STRING_TO_ARRAY(roles.permissions, ' ') && '{read_channel}'::TEXT[]
AND userid = 'jwedp8tbubn68mzbf3ufw1xczo';
```

The goal is to retrieve a list of channels that "have" the given permission. These are the pseudo steps:
* Get all channelmembers with the userid 'jwedp8tbubn68mzbf3ufw1xczo'.
* Unnest the roles (because they're stored as a string array) on channelmembers. This basically gives you something like a row per item in the string array so we can join the roles on them.
* Join the channels and roles tables on the results.
* Return only the channels that have the role 'read_channel' in the permissions of the role.

The MySQL query does the same thing just much less elegantly because it doesn't have the ability to "unnest" so you have to the ugly `select [N] union all` for each item in the roles array.

File: api4/post.go (hunk 4 above)
Comment (jwilander) [reply api4/post.go]: If I'm correct, `intersectChannelNames` will include all the names of channels on the team `teamID` that the user has the for `READ_CHANNEL` permission, right? So if `intersectChannelNames` is empty you're clearing the search parameters. That then gets passed down all the way to the posts store `Search()` function which will return an empty post list since search params is empty.

So that works well for the case of searching public and private channels, but this endpoint would also return results for direct/group message channels as well. DMs/GMs don't have team IDs associated with them though (it's just a blank string), so unless I'm reading the code wrong they aren't getting handled here. E.g. in a case where the user has no `READ_CHANNEL` permissions on any channel in a team, they /should/ still get search results returned for DMs/GMs but with this bit here that won't happen.

That make sense?

It think it might simply be removing this check, since `InChannel` should already be blank. You might also be able to check for blank team IDs and include those in a separate array, maybe `dmChannels`, and also check here if that array is empty

Comment (mkraft) [reply api4/post.go]: Yes you have read the code perfectly. 

However, `read_channel` is required by all channel types (open, private, direct, and group).

If for some reason the `ChannelMember` doesn't have a role that includes `read_channel` then it stands to reason that the results should not be returned, as per [ADR 1](https://gist.github.com/mkraft/17f62c66802bf2d68806586ecede97bd).

**Comment by mkraft** [comment]
Putting this PR into 'work in progress' because the SQL implementation will not be feasible with the implementation of "schemes" in phase 2 advanced permissions.

File: api4/post.go (hunk 4 above)
Comment (jwilander) [reply api4/post.go]: Right you should have `READ_CHANNEL` for every channel you can read but this will be skipping DMs/GMs because of them having a blank team ID, not because they're missing `READ_CHANNEL` (which they shouldn't be).

Comment (mkraft) [reply api4/post.go]: @jwilander You're right. I didn't realize DM's didn't have a team_id. Good catch. I'll add a test for that and make the appropriate change for the new implementation.
